### PR TITLE
Fix proxy UID/GID in Gateway templates

### PR DESCRIPTION
**Pull request by jwendell**

### CODE REVIEW

**Comment by jwendell** [comment]
Another approach to https://github.com/istio/istio/pull/46835

File: manifests/charts/gateway/templates/deployment.yaml [hunk 1]
```diff
@@ -65,9 +65,6 @@ spec:
             allowPrivilegeEscalation: false
             privileged: false
             readOnlyRootFilesystem: true
-            runAsUser: {{ .ProxyUID | default "1337" }}
```
Comment (howardjohn) [inline manifests/charts/gateway/templates/deployment.yaml:68]: I don't think this is equivilent, a lot of policies will deny higher level objects like Deployment early

Comment (jwendell) [reply manifests/charts/gateway/templates/deployment.yaml:68]: not sure I get it. This template is mixed with the one in the config map, right? So the end result should be the same?

Comment (howardjohn) [reply manifests/charts/gateway/templates/deployment.yaml:68]: The deployment is applied as-is, then the configmap is overlayed once the Pod is created.

But the deployment will never get modified

Comment (jwendell) [reply manifests/charts/gateway/templates/deployment.yaml:68]: got it. I modified it to use `.Capabilities` to check if this is an OpenShift cluster. Not sure if this is acceptable or not. It worked with `help install`. With `helm template` it will return false, falling back to including the block.

Comment (jwendell) [reply manifests/charts/gateway/templates/deployment.yaml:68]: It also doesn't work if we use the IstioOperator resource. It only works on `helm install`. This is fine for us.

Comment (howardjohn) [reply manifests/charts/gateway/templates/deployment.yaml:68]: Why don't we use platform=openshift like we do elsewhere?

If its to avoid "global", we should just put it as `.Values.platform.openshift` rather than `.Values.global.platform`

Comment (jwendell) [reply manifests/charts/gateway/templates/deployment.yaml:68]: @howardjohn done. Hopefully it's fine now.

File: manifests/charts/gateway/values.yaml [hunk 2]
```diff
@@ -139,3 +139,8 @@ volumes: []
 # A list of `VolumeMounts` added into the Gateway Pods. See
 # https://kubernetes.io/docs/concepts/storage/volumes/.
 volumeMounts: []
+
+# Platform where Istio is deployed. Possible values are: "openshift", "gcp".
+# An empty value means it is a vanilla Kubernetes distribution, therefore no special
+# treatment will be considered.
+platform: ""
```
Comment (bleggett) [inline manifests/charts/gateway/values.yaml:146]: Yeah again as mentioned previously in Slack - having a "platform" value that implies $some-arbitrary-set-of-other-values is far less obvious and flexible in the long term than just putting platform-specific values overrides a specific platform values file.

The values overrides a given platform needs should be 

1. All in a single file
2. Easy to add to or subtract from or alter *without* making template changes every time.

e.g. if say GCP needs a new override that Openshift already needs/requires or vice-versa - we shouldn't need to update a conditional tree in the templates for that. If we do we're sort of intentionally defeating how Helm and values overrides are designed to be used.

`openshift-values.yaml` > `--set values.platform=openshift`

**Review by howardjohn — APPROVED** [review]

**Review by bleggett — CHANGES_REQUESTED** [review]
Yeah again as mentioned previously in Slack - having a "platform" value that implies $some-arbitrary-set-of-other-values hidden within template conditionals is far less obvious and flexible in the long term than just putting platform-specific values overrides a specific platform values file.

This is not good Helm practice and leads to messy templates and nests of intersecting template conditionals.

The values overrides for a given branded platform should be 

1. All in a single file
2. Easy to add to or subtract from or alter *without* making template changes every time.

`platform` flags and template conditionals that query them can't satisfy this, and tend to create a mess.

e.g. if say GCP needs a new override that Openshift already needs/requires or vice-versa - we shouldn't need to update a conditional tree in the templates for that, and with this approach we will have to.

If we do `values.platform` we're sort of intentionally defeating how Helm and values overrides are designed to be used and creating future work for ourselves.

tl;dr:

`-f openshift-values.yaml` > `--set values.platform=openshift`

**Comment by jwendell** [comment]
ok, switched to a values overlay file.
